CPAC/pipeline/nodeblock.py

```diff
@@ -0,0 +1,115 @@
+from typing import Callable, List, Union, Optional, Dict, Any
+
+
+class NodeBlockFunction(object):
+    """
+    Stores a reference to the nodeblock function and all of its meta-data.
+    """
+
+    def __init__(
+            self,
+            func: Callable,
+            name: Optional[str] = None,
+            config: Optional[List[str]] = None,
+            switch: Optional[Union[List[str], List[List[str]]]] = None,
+            option_key: Optional[Union[str, List[str]]] = None,
+            option_val: Optional[Union[str, List[str]]] = None,
+            inputs: Optional[List[Union[str, list, tuple]]] = None,
+            outputs: Optional[Union[List[str], Dict[str, Any]]] = None
+    ) -> None:
+        self.func = func
+        """Nodeblock function reference."""
+        self.name: Optional[str] = name
+        """Used in the graph and logging to identify the NodeBlock and its component nodes."""
+        self.config: Optional[List[str]] = config
+        """
+        Indicates the nested keys in a C-PAC pipeline configuration should configure a NodeBlock built from this
+        function. If config is sct to "None", then all other configuration-related entities must be specified from the
+        root of the configuration.
+        """
+        self.switch: Optional[Union[List[str], List[List[str]]]] = switch
+        """
+        Indicates any keys that should evaluate to True for this NodeBlock to be active. A list of lists of strings
+        indicates multiple switches that must all be True to run, and is currently only an option if config is set to
+        "None".
+        """
+        self.option_key: Optional[Union[str, List[str]]] = option_key
+        """
+        Indicates the nested keys (starting at the nested key indicated by config) that should configure this NodeBlock.
+        """
+        self.option_val: Optional[Union[str, List[str]]] = option_val
+        """Indicates values for which this NodeBlock should be active."""
+        self.inputs: Optional[List[Union[str, list, tuple]]] = inputs
+        """ResourcePool keys indicating files needed for the NodeBlock's functionality."""
+        self.outputs: Optional[Union[List[str], Dict[str, Any]]] = outputs
+        """
+        ResourcePool keys indicating files generated or updated by the NodeBlock, optionally including metadata
+        for the outputs' respective sidecars.
+        """
+
```

Comment:
I think we should pass through the functions docstring, either raw like
```suggestion
        # pass docstring through
        self.__doc__ = func.__doc__

```
or including the old-style, pre-decorator dicts (using `pprint.pformat` in this suggestion):
```suggestion
        # pass docstring through + nodeblock dict
        self.__doc__ = "\n".join([
            func.__doc__, "Node Block:",
            pformat(self.legacy_nodeblock_dict(), sort_dicts=False)])

```

Reply:
I do prefer the first solution as I think there should not be "raw data" in the docstrings. but we might want to forward [all fields `functools.wraps` does](https://docs.python.org/3/library/functools.html#functools.update_wrapper)
What we could consider is converting the nodeblock data to more human readable documentation and include it. (E.g. "This nodeblock has the following inputs {...inputs}")

Reply:
Done in [51a8023](https://github.com/FCP-INDI/C-PAC/pull/1899/commits/51a8023e15fb47386b6786d1b478fab8ca6852db)